fix(reduce transform): surround invalid path segments with quotes (#21201)
vectordotdev/vector: changelog.d/21201_fields_iter.fix.md

> The `reduce` transform can now reduce fields that contain special characters.

Author: FYI @jszwedko I accepted your suggestion but then I realized this never paniced, it just failed to reduce when the path was not quoted.

> The fields iterator now quotes all invalid path segments and he reduce transform can correctly parse path containing special characters.

Reviewer: ```suggestion
The `reduce` transform no longer panics when reducing fields that contain special characters.
```

Nit: I think this will be a bit more understandable to Vector users. The bits about the iterator are only relevant to developers, who will have the commit logs.
